### PR TITLE
PCHR-1472: Prevent having more than one contract revision with the same effective date

### PR DESCRIPTION
**Pull request by omarabuhussein**
## Problem 

There cannot be more than one contract revision with the same effective date. 
So a validation  need to be added when creating a new revision to prevent that .

## Solution 

I've created a new API end-point for **HRJobContractRevision** entity called **validateeffectivedate** which is accept a contact_id and effective_date as parameters , the API end-point is an actually a wraper for a new static method **CRM_Hrjobcontract_BAO_HRJobContractRevision::validateEffectiveDate** which handle the validation logic. The API end-point is called via the job contract angular app after click on "Save Revision" and throw an error with the title of the contract that has conflicts with inputted effective date in case there is any :

anim

### CODE REVIEW

File: hrjobcontract/js/src/job-contract/services/contract-revision.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+             * @returns {*}
+             */
+            validateEffectiveDate: function(params) {
+                params.effective_date = convertToDateString(params.effective_date);
+                params.sequential = 0;
+                params.debug = settings.debug;
+
+                var deffered = $q.defer(), val;
+
+                ContractRevision.save({
```
Comment (danilo-compucorp) [inline hrjobcontract/js/src/job-contract/services/contract-revision.js]: You don't need to use `$q` here.

Just return the `$promise` from the `ContractRevision` resource.

**Comment by coldrunKacper** [comment]
Approved.

File: hrjobcontract/js/src/job-contract/services/contract-revision.js (hunk 1 above)
Comment (omarabuhussein) [inline hrjobcontract/js/src/job-contract/services/contract-revision.js]: I tired this :

```javascript
                return ContractRevision.save({ action: 'validateeffectivedate', json: params }).$promise
                    .then(function(result){
                        return result.values;
                    });
```
but in the case the Ajax request will not be triggered at all .

Comment (danilo-compucorp) [inline hrjobcontract/js/src/job-contract/services/contract-revision.js]: Please have a look at http://jsbin.com/qaxamas/edit?js,output

**Comment by danilo-compucorp** [comment]
Approved

**Pull request merged**